src/errors.rs

```diff
@@ -1,79 +1,85 @@
-                         --allow-prerelease flag might solve the issue."
-            )
-        }
-        /// Unable to parse external Cargo.toml
-        ParseCargoToml {
-            description("Unable to parse external Cargo.toml")
-        }
-        /// Cargo.toml could not be found.
-        MissingManifest {
-            description("Unable to find Cargo.toml")
-        }
-        /// Cargo.toml is valid toml, but doesn't contain the expected fields
-        InvalidManifest {
-            description("Cargo.toml missing expected `package` or `project` fields")
-        }
-        /// Found a workspace manifest when expecting a normal manifest
-        UnexpectedRootManifest {
-            description("Found virtual manifest, but this command requires running against an \
-                         actual package in this workspace.")
-        }
-        /// The TOML table could not be found.
-        NonExistentTable(table: String) {
-            description("non existent table")
-            display("The table `{}` could not be found.", table)
-        }
-        /// The dependency could not be found.
-        NonExistentDependency(name: String, table: String) {
-            description("non existent dependency")
-            display("The dependency `{}` could not be found in `{}`.", name, table)
-        }
-        /// Config of cargo is invalid
-        InvalidCargoConfig {
-            description("Invalid cargo config")
-        }
-        /// Unable to find the source specified by 'replace-with'
-        NoSuchSourceFound(name: String) {
-            description("Unable to find the source specified by 'replace-with'")
-            display("The source '{}' could not be found", name)
-        }
-        /// Unable to find the specified registry
-        NoSuchRegistryFound(name: String) {
-            display("The registry '{}' could not be found", name)
-        }
-        /// Failed to parse a version for a dependency
-        ParseVersion(version: String, dep: String) {
-            description("Failed to parse a version for a dependency")
-            display("The version `{}` for the dependency `{}` couldn't be parsed", version, dep)
-        }
-    }
+/// Main error type
+#[derive(Debug, ThisError)]
+pub enum Error {
+    // foreign links
+    /// An error from the std::io module
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+    ///  An error from the git2 crate
+    #[error(transparent)]
+    Git(#[from] git2::Error),
+    /// An error from the cargo_metadata crate
+    #[error(transparent)]
+    CargoMetadata(#[from] cargo_metadata::Error),
+    /// An error from the toml_edit crate
+    #[error(transparent)]
+    TomlEditParse(#[from] toml_edit::TomlError),
+    /// A ReqParseError from the semver crate
+    #[error(transparent)]
+    SemVerParse(#[from] semver::ReqParseError),
+    /// A SemVerError from the semver crate
+    #[error(transparent)]
+    SemVer(#[from] semver::SemVerError),
```

Comment:
can we merge these into one?

Reply:
Unfortunately source errors are completely separate types so we'd either lose those or have to cast them to trait objects or something